common-ui-compose/src/main/java/app/tivi/common/compose/Elevation.kt

```diff
@@ -0,0 +1,238 @@
+ */
+
+@file:Suppress("NOTHING_TO_INLINE")
+
+package app.tivi.common.compose
+
+import androidx.compose.foundation.AmbientIndication
+import androidx.compose.foundation.BorderStroke
+import androidx.compose.foundation.InteractionState
+import androidx.compose.foundation.ProvideTextStyle
+import androidx.compose.foundation.clickable
+import androidx.compose.foundation.indication
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.PaddingValues
+import androidx.compose.foundation.layout.Row
+import androidx.compose.foundation.layout.RowScope
+import androidx.compose.foundation.layout.defaultMinSizeConstraints
+import androidx.compose.foundation.layout.padding
+import androidx.compose.material.ButtonColors
+import androidx.compose.material.ButtonConstants
+import androidx.compose.material.ButtonElevation
+import androidx.compose.material.ElevationOverlay
+import androidx.compose.material.ExperimentalMaterialApi
+import androidx.compose.material.MaterialTheme
+import androidx.compose.material.Surface
+import androidx.compose.material.contentColorFor
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.Providers
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.staticAmbientOf
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.graphics.RectangleShape
+import androidx.compose.ui.graphics.Shape
+import androidx.compose.ui.graphics.compositeOver
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+import kotlin.math.ln
+
+val AmbientAbsoluteElevation = staticAmbientOf { 0.dp }
+
+/**
+ * A wrapper around [Surface] which works with absolute elevation. This will automatically
+ * update [AmbientAbsoluteElevation], adding [elevation].
+ */
+@Composable
+fun AbsoluteElevationSurface(
+    modifier: Modifier = Modifier,
+    shape: Shape = RectangleShape,
+    color: Color = MaterialTheme.colors.surface,
+    contentColor: Color = contentColorFor(color),
+    border: BorderStroke? = null,
+    elevation: Dp = 0.dp,
+    content: @Composable () -> Unit
+) {
+    val currentAbsElevation = AmbientAbsoluteElevation.current
+    Surface(
+        modifier = modifier,
+        shape = shape,
+        color = color,
+        contentColor = contentColor,
+        border = border,
+        elevation = elevation
+    ) {
+        Providers(AmbientAbsoluteElevation provides currentAbsElevation + elevation) {
+            content()
+        }
+    }
+}
+
+/**
+ * An [ElevationOverlay] which supports absolute elevation, taking the current
+ * [AmbientAbsoluteElevation] into account.
+ */
+object AbsoluteElevationOverlay : ElevationOverlay {
+    @Composable
+    override fun apply(color: Color, elevation: Dp): Color {
+        val colors = MaterialTheme.colors
+        val absElevation = AmbientAbsoluteElevation.current + elevation
```

Comment:
Is this "double counting" e.g. if an AbsoluteElevationSurface has an elevation that it passes through to this, hasn't this already been added to the absolute value?

Reply:
Ack, yeah a comment would help as I wasn't _sure_ if it was double counting or not lol.